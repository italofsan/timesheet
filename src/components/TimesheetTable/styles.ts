import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import { colors } from "../../styles/colors";

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    tableCellHead: {
      backgroundColor: colors.orange,
      color: colors.white,
    },
    tableCellHeadOdd: {
      backgroundColor: colors.main,
      color: colors.white,
    },
    optionsButton: {
      margin: "0 8px",
    },
  })
);
