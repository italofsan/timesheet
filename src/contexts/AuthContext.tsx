import {
  createContext,
  Dispatch,
  ReactNode,
  SetStateAction,
  useEffect,
  useState,
} from "react";
import { auth, firebase } from "../services/firebase";
import { errorMessage } from "../components/Messages";

type User = {
  id: string;
  name: string;
};

type AuthContextType = {
  user: User;
  signInWithGoogle: () => Promise<void>;
  signOut: () => Promise<void>;
  isSigned: boolean;
  setIsSigned: Dispatch<SetStateAction<boolean>>;
};

type AuthContextProviderProps = {
  children: ReactNode;
};

export const AuthContext = createContext({} as AuthContextType);

export function AuthContextProvider(props: AuthContextProviderProps) {
  const [user, setUser] = useState<User>({
    id: "",
    name: "",
  });
  const [isSigned, setIsSigned] = useState<boolean>(false);

  function loadStorageData() {
    const storageIsSigned = localStorage.getItem("@SIGNED");

    if (storageIsSigned) {
      setIsSigned(JSON.parse(storageIsSigned));
    }
  }

  useEffect(() => {
    loadStorageData();
    const unsubscribe = auth.onAuthStateChanged((user) => {
      if (user) {
        const { displayName, uid } = user;

        if (!displayName) {
          throw new Error("Missing information from Google Account.");
        }

        setUser({
          id: uid,
          name: displayName,
        });
        setIsSigned(true);
        localStorage.setItem("@SIGNED", JSON.stringify(true));
      }
    });

    return () => {
      unsubscribe();
    };
  }, []);

  async function signInWithGoogle() {
    const provider = new firebase.auth.GoogleAuthProvider();
    try {
      const result = await auth.signInWithPopup(provider);
      if (result.user) {
        const { displayName, uid } = result.user;
        if (!displayName) {
          throw new Error("Missing information from Google Account.");
        }
        setUser({
          id: uid,
          name: displayName,
        });
      }
    } catch (error) {
      console.log(error);
      errorMessage("Não foi possível realizar Login!");
    }
  }

  async function signOut() {
    await auth.signOut();
    setUser({} as User);
    localStorage.clear();
    setIsSigned(false);
  }

  return (
    <AuthContext.Provider
      value={{
        user,
        signInWithGoogle,
        signOut,
        isSigned,
        setIsSigned,
      }}
    >
      {props.children}
    </AuthContext.Provider>
  );
}
