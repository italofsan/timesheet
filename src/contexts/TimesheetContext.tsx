import {
  createContext,
  ReactNode,
  useCallback,
  useEffect,
  useState,
} from "react";
import { errorMessage, successMessage } from "../components/Messages";
import { DataTimesheetInput, DataTimesheet } from "../utils/types";
import {
  calcUndertimeHours,
  calcOvertimeHours,
  calcWorkedHours,
  sumHours,
} from "../utils/functions/mathHoursFunctions";
import { isTimesheetExists } from "../utils/functions/timesheetFunctions";
import { database } from "../services/firebase";
import { useAuth } from "../hooks/useAuth";

interface TimesheetContextProps {
  listTimesheets: DataTimesheet[];
  totalWorkedHours: string;
  totalOvertimeHours: string;
  totalUndertimeHours: string;
  handleAddTimesheet: (valuesForm: DataTimesheetInput) => void;
  handleUpdateTimesheet: (valuesForm: DataTimesheet) => void;
  handleDeleteTimesheet: (selectedTimesheetId: string) => void;
}

interface TimesheetContextProviderProps {
  children: ReactNode;
}

type FirebaseTimesheets = Record<string, DataTimesheet>;

export const TimesheetContext = createContext<TimesheetContextProps>(
  {} as TimesheetContextProps
);

export function TimesheetContextProvider({
  children,
}: TimesheetContextProviderProps) {
  const { user } = useAuth();
  const [listTimesheets, setListTimesheets] = useState<DataTimesheet[]>([]);
  const [totalWorkedHours, setTotalWorkedHours] = useState("00:00");
  const [totalOvertimeHours, setTotalOvertimeHours] = useState("00:00");
  const [totalUndertimeHours, setTotalUndertimeHours] = useState("00:00");

  useEffect(() => {
    setTotalWorkedHours(
      listTimesheets.reduce((sumTotal, log) => {
        return sumHours(sumTotal, log.workedHours);
      }, "00:00")
    );

    setTotalOvertimeHours(
      listTimesheets.reduce((sumTotal, log) => {
        return sumHours(sumTotal, log.overtimeHours);
      }, "00:00")
    );

    setTotalUndertimeHours(
      listTimesheets.reduce((sumTotal, log) => {
        return sumHours(sumTotal, log.undertimeHours);
      }, "00:00")
    );
  }, [listTimesheets]);

  useEffect(() => {
    const timesheetRef = database.ref();
    timesheetRef.on("value", (timesheet) => {
      const databaseTimesheet = timesheet.val();
      const firebaseTimesheets: FirebaseTimesheets = databaseTimesheet
        ? databaseTimesheet.timesheets
        : {};

      if (firebaseTimesheets) {
        const parsedTimesheets = Object.entries(firebaseTimesheets)
          .map(([key, value]) => {
            return {
              id: key,
              date: value.date,
              idUser: value.idUser,
              arrivalTime: value.arrivalTime,
              lunchExitTime: value.lunchExitTime,
              lunchArrivalTime: value.lunchArrivalTime,
              exitTime: value.exitTime,
              undertimeHours: value.undertimeHours,
              overtimeHours: value.overtimeHours,
              workedHours: value.workedHours,
              createdAt: value.createdAt,
            };
          })
          .filter((timesheet) => timesheet.idUser === user.id);

        setListTimesheets(parsedTimesheets);
      }
    });

    return () => {
      timesheetRef.off("value");
    };
  }, [user.id]);

  const handleAddTimesheet = useCallback(
    async (valuesForm: DataTimesheetInput) => {
      if (isTimesheetExists(listTimesheets, valuesForm)) {
        return errorMessage("You already have a register for this day!");
      }
      const timesheetRef = database.ref("timesheets");
      try {
        await timesheetRef.push({
          ...valuesForm,
          overtimeHours: calcOvertimeHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          undertimeHours: calcUndertimeHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          workedHours: calcWorkedHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          createdAt: String(new Date()),
          date: String(new Date(valuesForm.date)),
          idUser: user.id,
        });
        successMessage("Register created gracefully!");
      } catch (error) {
        errorMessage("Unable to create register!");
      }
    },
    [user.id, listTimesheets]
  );

  const handleUpdateTimesheet = useCallback(
    async (valuesForm: DataTimesheet) => {
      try {
        await database.ref(`timesheets/${valuesForm.id}`).update({
          arrivalTime: valuesForm.arrivalTime,
          lunchExitTime: valuesForm.lunchExitTime,
          lunchArrivalTime: valuesForm.lunchArrivalTime,
          exitTime: valuesForm.exitTime,
          overtimeHours: calcOvertimeHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          undertimeHours: calcUndertimeHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          workedHours: calcWorkedHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
        });
        successMessage("Register updated gracefully!");
      } catch (error) {
        errorMessage("Unable to update register!");
      }
    },
    []
  );

  const handleDeleteTimesheet = useCallback(
    async (selectedTimesheetId: string) => {
      try {
        await database.ref(`timesheets/${selectedTimesheetId}`).remove();
        successMessage("Register removed gracefully!");
      } catch (error) {
        errorMessage("Unable to remove register!");
      }
    },
    []
  );

  return (
    <TimesheetContext.Provider
      value={{
        listTimesheets,
        totalWorkedHours,
        totalOvertimeHours,
        totalUndertimeHours,
        handleAddTimesheet,
        handleUpdateTimesheet,
        handleDeleteTimesheet,
      }}
    >
      {children}
    </TimesheetContext.Provider>
  );
}
