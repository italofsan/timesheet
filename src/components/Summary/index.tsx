import { Grid, Paper, Typography } from "@material-ui/core";
import ArrowUpwardIcon from "@material-ui/icons/ArrowUpward";
import ArrowDownwardIcon from "@material-ui/icons/ArrowDownward";
import HistoryIcon from "@material-ui/icons/History";

import { useTimesheet } from "../../hooks/useTimesheet";

import { subtractHours } from "../../utils/functions/mathHoursFunctions";

import { useStyles } from "./styles";

export function Summary() {
  const classes = useStyles();
  const { totalWorkedHours, totalOvertimeHours, totalUndertimeHours } =
    useTimesheet();

  return (
    <>
      <Grid item xs={12} lg={4}>
        <Paper className={classes.workedHoursSummaryContainer}>
          <Typography className={classes.workedHoursSummaryValue}>
            {totalWorkedHours}
          </Typography>
          <Typography className={classes.workedHoursSummaryTitle}>
            Total Worked Hours
          </Typography>
          <HistoryIcon className={classes.arrowIcon} />
        </Paper>
      </Grid>
      <Grid item xs={6} lg={4}>
        <Paper className={classes.summaryContainer}>
          <Typography className={classes.overtimeHoursSummaryValue}>
            {subtractHours(totalUndertimeHours, totalOvertimeHours).length === 6
              ? "00:00"
              : `${subtractHours(totalUndertimeHours, totalOvertimeHours)}`}
          </Typography>
          <Typography className={classes.summaryTitle}>
            Total Overtime Hours
          </Typography>
          <ArrowUpwardIcon className={classes.arrowIcon} />
        </Paper>
      </Grid>
      <Grid item xs={6} lg={4}>
        <Paper className={classes.summaryContainer}>
          <Typography className={classes.undertimeHoursSummaryValue}>
            {subtractHours(totalOvertimeHours, totalUndertimeHours).length >= 6
              ? "00:00"
              : `${subtractHours(totalOvertimeHours, totalUndertimeHours)}`}
          </Typography>
          <Typography className={classes.summaryTitle}>
            Total Undertime Hours
          </Typography>
          <ArrowDownwardIcon className={classes.arrowIcon} />
        </Paper>
      </Grid>
    </>
  );
}
