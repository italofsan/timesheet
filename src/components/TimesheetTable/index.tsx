import { useState } from "react";
import {
  Grid,
  IconButton,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Tooltip,
  Zoom,
} from "@material-ui/core";
import { Edit, Delete } from "@material-ui/icons";

import { UpdateModal } from "../../components/Modals/UpdateModal";
import { useTimesheet } from "../../hooks/useTimesheet";

import { DataTimesheet } from "../../utils/types";

import { useStyles } from "./styles";
import { DeleteModal } from "../Modals/DeleteModal";
import { colors } from "../../styles/colors";

const columnsTimesheetTable = [
  "Options",
  "Date",
  "Arrival",
  "Lunch Exit",
  "Lunch Arrival",
  "Exit",
  "Worked Hours",
  "Overtime",
  "Undertime",
];

export function TimesheetTable() {
  const classes = useStyles();
  const { listTimesheets } = useTimesheet();

  const [selectedTimesheet, setSelectedTimesheet] = useState<DataTimesheet>(
    {} as DataTimesheet
  );

  const [updateModalOpen, setUpdateModalOpen] = useState(false);
  function handleOpenUpdateModal() {
    setUpdateModalOpen(true);
  }
  function handleCloseUpdateModal() {
    setUpdateModalOpen(false);
  }
  const [deleteModalOpen, setDeleteModalOpen] = useState(false);
  function handleOpenDeleteModal() {
    setDeleteModalOpen(true);
  }
  function handleCloseDeleteModal() {
    setDeleteModalOpen(false);
  }

  return (
    <Grid item xs={12}>
      <TableContainer component={Paper}>
        <Table>
          <caption
            style={{ backgroundColor: colors.main, color: "#FFF" }}
          >{`You have ${
            listTimesheets && listTimesheets.length !== 0
              ? listTimesheets.length
              : 0
          }  register${
            listTimesheets && listTimesheets.length !== 1 ? "s" : ""
          }!`}</caption>
          <TableHead>
            <TableRow>
              {columnsTimesheetTable.map((columnTimesheetTable, index) => (
                <TableCell
                  key={index}
                  align='center'
                  className={classes.tableCellHeadOdd}
                >
                  {columnTimesheetTable}
                </TableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {listTimesheets.map((timesheet, index) => (
              <TableRow key={index}>
                <TableCell align='center'>
                  <Tooltip
                    arrow
                    title='Edit'
                    placement='top'
                    enterDelay={500}
                    TransitionComponent={Zoom}
                  >
                    <IconButton
                      edge='start'
                      className={classes.optionsButton}
                      color='inherit'
                      aria-label='edit'
                      onClick={() => {
                        setSelectedTimesheet(timesheet);
                        handleOpenUpdateModal();
                      }}
                    >
                      <Edit />
                    </IconButton>
                  </Tooltip>
                  <Tooltip
                    arrow
                    title='Delete'
                    placement='top'
                    enterDelay={500}
                    TransitionComponent={Zoom}
                  >
                    <IconButton
                      edge='start'
                      className={classes.optionsButton}
                      color='inherit'
                      aria-label='delete'
                      onClick={() => {
                        setSelectedTimesheet(timesheet);
                        handleOpenDeleteModal();
                      }}
                    >
                      <Delete />
                    </IconButton>
                  </Tooltip>
                </TableCell>
                <TableCell align='center' style={{ backgroundColor: "#eef" }}>
                  {timesheet.date &&
                    new Intl.DateTimeFormat("en-US").format(
                      new Date(timesheet.date)
                    )}
                </TableCell>
                <TableCell align='center'>{timesheet.arrivalTime}</TableCell>
                <TableCell align='center' style={{ backgroundColor: "#eef" }}>
                  {timesheet.lunchExitTime}
                </TableCell>
                <TableCell align='center'>
                  {timesheet.lunchArrivalTime}
                </TableCell>
                <TableCell align='center' style={{ backgroundColor: "#eef" }}>
                  {timesheet.exitTime}
                </TableCell>
                <TableCell align='center'>{timesheet.workedHours}</TableCell>
                <TableCell align='center' style={{ backgroundColor: "#eef" }}>
                  {timesheet.overtimeHours}
                </TableCell>
                <TableCell align='center'>{timesheet.undertimeHours}</TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
      <UpdateModal
        selectedTimesheet={selectedTimesheet}
        updateModalOpen={updateModalOpen}
        handleCloseUpdateModal={handleCloseUpdateModal}
      />
      <DeleteModal
        selectedTimesheet={selectedTimesheet}
        deleteModalOpen={deleteModalOpen}
        handleCloseDeleteModal={handleCloseDeleteModal}
      />
    </Grid>
  );
}
