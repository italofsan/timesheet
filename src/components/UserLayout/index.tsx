import { ReactNode } from "react";
import {
  AppBar,
  Button,
  Container,
  Grid,
  Toolbar,
  Typography,
} from "@material-ui/core";

import { useHistory } from "react-router-dom";

import { useStyles } from "./styles";
import { useAuth } from "../../hooks/useAuth";

interface UserLayoutProps {
  children: ReactNode;
}

const UserLayout = ({ children }: UserLayoutProps) => {
  const classes = useStyles();
  const history = useHistory();
  const { signOut } = useAuth();

  function handleSignOut() {
    signOut();
    history.push("/");
  }

  return (
    <div className={classes.root}>
      <AppBar position='fixed' className={classes.appBar}>
        <Container maxWidth='lg'>
          <Toolbar>
            <Typography variant='h6' className={classes.title}>
              Timesheet App
            </Typography>
            <Button color='inherit' onClick={handleSignOut}>
              Logout
            </Button>
          </Toolbar>
        </Container>
      </AppBar>
      <Grid item xs={12}>
        <main className={classes.content}>
          <Toolbar />
          {children}
        </main>
      </Grid>
    </div>
  );
};

export default UserLayout;
