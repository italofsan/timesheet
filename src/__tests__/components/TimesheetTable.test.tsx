import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import { TimesheetContextProvider } from "../../contexts/TimesheetContext";
import { AuthContextProvider } from "../../contexts/AuthContext";

import { TimesheetTable } from "../../components/TimesheetTable";

describe("Timesheet Component", () => {
  it("renders correctly", () => {
    render(
      <AuthContextProvider>
        <TimesheetContextProvider>
          <TimesheetTable />
        </TimesheetContextProvider>
      </AuthContextProvider>
    );
    expect(screen.getByText("Date")).toBeInTheDocument();
  });

  const tree = renderer
    .create(
      <AuthContextProvider>
        <TimesheetContextProvider>
          <TimesheetTable />
        </TimesheetContextProvider>
      </AuthContextProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
