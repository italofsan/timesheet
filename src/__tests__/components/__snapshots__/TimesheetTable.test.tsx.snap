// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
>
  <div
    className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      className="MuiTable-root"
      role={null}
    >
      <caption
        style={
          Object {
            "backgroundColor": "rgb(143,125,238)",
            "color": "#FFF",
          }
        }
      >
        You have 0  registers!
      </caption>
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-2 MuiTableCell-alignCenter"
            scope="col"
          >
            Options
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-2 MuiTableCell-alignCenter"
            scope="col"
          >
            Date
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-2 MuiTableCell-alignCenter"
            scope="col"
          >
            Arrival
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-2 MuiTableCell-alignCenter"
            scope="col"
          >
            Lunch Exit
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-2 MuiTableCell-alignCenter"
            scope="col"
          >
            Lunch Arrival
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-2 MuiTableCell-alignCenter"
            scope="col"
          >
            Exit
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-2 MuiTableCell-alignCenter"
            scope="col"
          >
            Worked Hours
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-2 MuiTableCell-alignCenter"
            scope="col"
          >
            Overtime
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-2 MuiTableCell-alignCenter"
            scope="col"
          >
            Undertime
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      />
    </table>
  </div>
</div>
`;
